packages/viewport/src/with-viewport-match.js

```diff
@@ -33,11 +103,33 @@ import { withSelect } from '@wordpress/data';
  * @return {Function} Higher-order component.
  */
 const withViewportMatch = ( queries ) => createHigherOrderComponent(
-	withSelect( ( select ) => {
-		return mapValues( queries, ( query ) => {
-			return select( 'core/viewport' ).isViewportMatch( query );
-		} );
-	} ),
+	( WrappedComponent ) => {
+		return ( props ) => {
+			const globalQueriesResult = useSelect( ( select ) => {
+				return mapValues( queries, ( query ) => {
+					return select( 'core/viewport' ).isViewportMatch( query );
+				} );
+			}, [ queries ] );
+			const localQueries = useContext( ViewportMatchWidthContext );
+			const localQueriesResult = useMemo(
+				() => {
+					if ( ! localQueries ) {
+						return null;
+					}
+					return mapValues( queries, ( query ) => {
+						return localQueries[ query ];
+					} );
+				},
+				[ localQueries, queries ]
+			);
+			return (
+				<WrappedComponent
+					{ ...props }
+					{ ...( localQueriesResult ? localQueriesResult : globalQueriesResult ) }
```

Comment:
I'm not sure if there are options, but it seems wasteful to compute _both_ the local and global results, only to ultimately pass just one along.

The only thing I could consider is if we extracted separate subcomponents for each whether there is or isn't context, but I suspect that extra layer might just make things more convoluted anyways.

```jsx
function GlobalViewportResult( { props } ) {}

function ContextViewportResult( { props } ) {}

return ( props ) => {
	const width = useContext( ViewportMatchWidthContext );
	return width ? 
		<ContextViewportResult { ...props } width={ width } /> : 
		<GlobalViewportResult { ...props } />;
};
```

🤷‍♂